duet/impl.py

```diff
@@ -258,6 +319,28 @@ def spawn(self, awaitable: Awaitable[Any], main_task: Optional[Task] = None) ->
         self._ready_tasks.register(task)
         return task
 
+    def time(self) -> float:
+        return time.time()
+
+    def add_deadline(self, task: Task, deadline: float) -> DeadlineEntry:
+        entry = DeadlineEntry(task, deadline=deadline)
+        heapq.heappush(self._deadlines, entry)
+        return entry
+
+    def get_next_deadline(self) -> Optional[float]:
+        while self._deadlines:
+            if not self._deadlines[0].valid:
+                heapq.heappop(self._deadlines)
+                continue
```

Comment:
I apologize if I'm asking obvious questions, but I'm a bit confused about some of the logic -- in order for the deadline entry to not be valid, we'd have to have called `pop_deadline`, right? That means the scope has exited, I think. If I'm following this, it seems like either this code shouldn't happen, or we should interrupt the task with a timeout error?

Reply:
The idea here is that when the task exits a scope before the deadline, we no longer need to do anything when that deadline eventually elapses. We can easily pop it off the task's deadline stack, but we don't know the index of that entry in the scheduler's deadline queue so we can't remove it from the scheduler queue without doing a linear search. Instead, we mark the entry as invalid but leave it in the queue. When it eventually elapses, we get to this code which pops the invalid deadline off the queue and moves on to the next entry.

Reply:
Oh I see! What you wrote is perfect, and would make for a helpful comment by the `valid` attribute definition, something like `valid indicates that the task has completed and the deadline no longer needs to be considered. We don't bother removing stale deadline entries for performance reasons, and instead ignore invalid entries` or something like that?

Reply:
Reworked this as an `Attributes` section in the docstring.

Reply:
Thank you for adding that! It super helps readability :)